client-js/app/scripts/obligation.ts

```diff
@@ -0,0 +1,48 @@
+/**
+ * Obligation is the input side complement of Promise; if one party is promised
+ * a value, then of necessity another party must be obliged to provide the
+ * value or explain why they failed to do so. Obligation has two methods
+ * `resolve` and `reject` which can be used to send a value or an error into a
+ * chain of asynchronous callbacks. Promise is the output side of the chain,
+ * providing access to the asynchronous values by adding callback functions.
+ *
+ * In the traditional ES6 Promise API, resolve and reject functions are only
+ * available as arguments passed in to the executor function we pass to the
+ * Promise constructor. But in many situations we need to store the callbacks
+ * for later use or access them outside the constructor, which Obligation makes
+ * possible. This is especially useful in situations where we need to interface
+ * traditional callback-based code with Promise-based code.
+ *
+ * Compare the scala.concurrent library, which properly separates the input and
+ * output sides of an asynchronous callback chain into two classes:
+ * scala.concurrent.Promise is the input side (equivalent to Obligation here)
+ * and scala.concurrent.Future is the output side (equivalent to Promise here).
+ */
+export class Obligation<R> {
+
+  public promise: Promise<R>;
```

Comment:
I have read and comprehended the code in https://github.com/labrad/scalabrad-web/commit/2b79ab5ddeabc3e9afe302f7d032a981eea21a86, and I like it much better.
Having the `Promise` live inside an `Obligation` is asymmetric for no reason. In the new style where we say
```javascript
var {obligation, promise} = obligate<A>();
```
the symmetry makes sense; `obligate` and `promise` are on equal footing.